superset-frontend/src/filters/components/Select/SelectFilterPlugin.tsx

```diff
@@ -18,31 +18,87 @@
  */
 import {
   AppSection,
+  DataMask,
   ensureIsArray,
+  ExtraFormData,
   GenericDataType,
+  JsonObject,
   smartDateDetailedFormatter,
   t,
   tn,
 } from '@superset-ui/core';
-import React, { useEffect, useState } from 'react';
+import React, { useEffect, useReducer, useState } from 'react';
 import { Select } from 'src/common/components';
-import { FIRST_VALUE, PluginFilterSelectProps, SelectValue } from './types';
+import { debounce } from 'lodash';
+import { SLOW_DEBOUNCE } from 'src/constants';
+import { PluginFilterSelectProps, SelectValue } from './types';
 import { StyledSelect, Styles } from '../common';
 import { getDataRecordFormatter, getSelectExtraFormData } from '../../utils';
 
 const { Option } = Select;
 
+const debouncedOwnStateFunc = debounce(
+  (dispatch: { (action: DataMaskAction): void }, val: string) => {
+    dispatch({
+      type: 'ownState',
+      ownState: {
+        search: val,
+      },
+    });
+  },
+  SLOW_DEBOUNCE,
+);
+
+type DataMaskAction =
+  | { type: 'ownState'; ownState: JsonObject }
+  | {
+      type: 'filterState';
+      extraFormData: ExtraFormData;
+      filterState: { value: SelectValue };
+    };
+
+function reducer(state: DataMask, action: DataMaskAction): DataMask {
+  switch (action.type) {
+    case 'ownState':
+      return {
+        ...state,
```

Comment:
may be try to use here Immer, it should reduce code

Reply:
I'll refactor in a follow-up PR if that's ok (we might want to look at some of the linter settings if we want to make this the preferred approach)